src/Core/Managed/Shared/Extensibility/MetricAggregatorManager.cs

```diff
@@ -0,0 +1,295 @@
+            // to make perceived system latency look smaller
+            var nextWakeTime = DateTimeOffset.MinValue
+                .AddMinutes((long)minutesFromZero)
+                .Add(snapshotFrequency)
+                .AddSeconds(1);
+
+            TimeSpan sleepTime = nextWakeTime - DateTimeOffset.UtcNow;
+
+            // adjust wait time to a bit longer than a minute if the wake up time is within few seconds from now
+            return sleepTime < TimeSpan.FromSeconds(3) ? sleepTime.Add(snapshotFrequency) : sleepTime;
+        }
+
+        /// <summary>
+        /// Generates telemetry object based on the metric aggregator.
+        /// </summary>
+        /// <param name="aggregator">Metric aggregator.</param>
+        /// <param name="aggregatorStats">Metric aggregator statistics calculated for a period of time.</param>
+        /// <returns>Metric telemetry object resulting from aggregation.</returns>
+        private static AggregatedMetricTelemetry CreateAggergatedMetricTelemetry(MetricAggregator aggregator, SimpleMetricStatisticsAggregator aggregatorStats)
+        {
+            if ((aggregator == null) || (aggregatorStats == null) || (aggregatorStats.Count <= 0))
+            {
+                return null;
+            }
+
+            var telemetry = new AggregatedMetricTelemetry(
+                aggregator.MetricName,
+                aggregatorStats.Count,
+                aggregatorStats.Sum,
+                aggregatorStats.Min,
+                aggregatorStats.Max,
+                aggregatorStats.StandardDeviation);
+
+            if (aggregator.Dimensions != null)
+            {
+                foreach (KeyValuePair<string, string> property in aggregator.Dimensions)
+                {
+                    telemetry.Properties.Add(property);
+                }
+            }
+
+            return telemetry;
+        }
+
+        private void Flush(bool disposing)
+        {
+            if ((!disposing) && this.disposed)
+            {
+                throw new ObjectDisposedException(nameof(MetricAggregatorManager));
+            }
+
+            try
+            {
+                this.Snapshot();
+                this.telemetryClient.Flush();
+            }
+            catch (Exception ex)
+            {
+                CoreEventSource.Log.FailedToFlushMetricAggregators(ex.ToString());
+            }
+        }
+
+        /// <summary>
+        /// Takes a snapshot of aggregators collected by this instance of the manager
+        /// and schedules the next snapshot.
+        /// </summary>
+        private Task SnapshotAndReschedule()
+        {
+            return Task.Factory.StartNew(
+                () =>
+                {
+                    try
+                    {
+                        this.Snapshot();
+                    }
+                    catch (Exception ex)
+                    {
+                        CoreEventSource.Log.FailedToSnapshotMetricAggregators(ex.ToString());
```

Comment:
I'd rather have a specific message logged here in addition to more generic "timed task failed" message logged by timed task.